core/meta/src/TClass.cxx

```diff
@@ -3678,20 +3679,46 @@ TList *TClass::GetListOfDataMembers(Bool_t load /* = kTRUE */)
             // R__ASSERT(kFALSE);
 
             fHasRootPcmInfo = kTRUE;
-            return fData;
+            return data;
          }
       }
-      fData = new TListOfDataMembers(this);
+
+      data = new TListOfDataMembers(this, selection);
    }
-   if (Property() & (kIsClass|kIsStruct|kIsUnion)) {
+   if (IsClassStructOrUnion()) {
       // If the we have a class or struct or union, the order
       // of data members is the list is essential since it determines their
       // order on file.  So we must always load.  Also, the list is fixed
       // since the language does not allow to add members.
-      if (!fData->IsLoaded()) fData->Load();
+      if (!data->IsLoaded())
+         data->Load();
+
+   } else if (load) data->Load();
+   return data;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Return list containing the TDataMembers of a class.
+
+TList *TClass::GetListOfDataMembers(Bool_t load /* = kTRUE */)
+{
+   // Fast path, no lock? Classes load at creation time.
+   if (fData && (!load || IsClassStructOrUnion()))
```

Comment:
To skip taking the lock here, `fData` must become an `std::atomic` otherwise it can be read and return 'partially' while another thread is updating.